server/core/include/scoped_client_identity.hpp

```diff
@@ -0,0 +1,56 @@
+#ifndef IRODS_SCOPED_CLIENT_IDENTITY_HPP
+#define IRODS_SCOPED_CLIENT_IDENTITY_HPP
+
+#include "rcConnect.h"
+
+#include <cstring>
+#include <string>
+#include <string_view>
+#include <stdexcept>
+
+namespace irods::experimental
+{
+    /// This class provides a convenient RAII-style mechanism for becoming a different
+    /// iRODS user for the duration of a scoped block.
+    ///
+    /// Instances of this class are not copyable or moveable.
+    ///
+    /// \user Server
+    ///
+    /// \since 4.2.8
+    class scoped_client_identity
+    {
+    public:
+        /// Constructs a scoped_client_identity object and sets the username of
+        /// \b comm to \b new_username.
+        ///
+        /// \param[in] comm         The server communication object.
+        /// \param[in] new_username The name of the iRODS user to become.
+        scoped_client_identity(rsComm_t& comm, std::string_view new_username)
+            : conn_{comm}
+            , username_{conn_.clientUser.userName}
```

Comment:
If an exception is thrown from the constructor, then we lose cycles by doing this step. However, the cost of doing this is likely very small and the implementation is correct as is.